EIPS/eip-7791.md

```diff
@@ -0,0 +1,82 @@
+---
+eip: 7791
+title: GAS2ETH opcode
+description: Introduces a new opcode, `GAS2ETH`, to convert gas/mana to ETH
+author: Charles Cooper (@charles-cooper), Pascal Caversaccio (@pcaversaccio)
+discussions-to: https://ethereum-magicians.org/t/eip-7791-gas2eth-opcode/21418
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-08-13
+requires: 2929
+---
+
+## Abstract
+
+This EIP introduces a new `GAS2ETH` opcode that enables the direct conversion of gas/mana into ether (ETH).
+
+## Motivation
+
+This EIP is based on the premise that smart contract authors, compiler teams and public goods projects in general should be compensated for their contributions.
+Moreover, their compensation should scale with the usage of their contracts.
+A widely used and popular contract offers significant value to its users through its functionality and to the network by driving demand for blockspace — Ethereum's _raison d'être_.
+This increased demand also benefits miners and validators, who are rewarded for executing these contracts.
+
+Monetizing smart contracts in a scalable manner remains challenging at the time of this writing.
+This difficulty is evident from existence of many different monetization strategies employed across various smart contracts — ranging from fee structures to the issuance of tokens with "tokenomics" of varying levels of complexity.
+Additionally, many public goods projects struggle to secure funding.
+
+Introducing the `GAS2ETH` opcode offers contract authors, as well as the tools they use, a new way to achieve their monetization objectives.
+By charging gas, they integrate with an established user experience that is both familiar and understood by users.
+The proposed instruction ensures that existing transaction creation and processing tools remain unchanged.
+Moreover, by charging gas, contract authors align economically with network activity; they benefit from higher compensation during periods of intense network usage and receive less when activity is low.
+This helps align the incentives of smart contract authors, validators, and the broader network.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+A new opcode is introduced, `GAS2ETH` (`0xFC`), which:
+
+- Pops two values from the stack: `addr` then `gas_amount`. If there are fewer than two values on the stack, the calling context should fail with stack underflow.
+- Deducts `gas_amount` from the current calling context.
+- Computes `wei_val` by multiplying `gas_amount` by the current transaction context's `gas_price`.
```

Comment:
What is the gas price? Is this the base fee? Or is this the tip? Or is it both?

Reply:
interesting question, i think it should be the tx gas price which is used to calculate the tx cost in ETH, a la https://github.com/ethereum/execution-specs/blob/98d6ddaaa709a2b7d0cd642f4cfcdadc8c0808e1/src/ethereum/berlin/fork.py#L680.